flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/codegen/HashCodeGeneratorTest.scala

```diff
@@ -41,16 +45,164 @@ class HashCodeGeneratorTest {
+    )
+    assertEquals(1065781729, hashFunc3.hashCode(row3))
+
+    // test hash code for ArrayData
+    CommonTestUtils.assertThrows(
+      "RowData hash function doesn't support to generate hash code for ArrayData.",
+      classOf[RuntimeException],
+      () => hashFunc3.hashCode(new GenericArrayData(Array(1)))
+    )
+
+    // test hash code for MapData
+    CommonTestUtils.assertThrows(
+      "RowData hash function doesn't support to generate hash code for MapData.",
+      classOf[RuntimeException],
+      () => hashFunc3.hashCode(new GenericMapData(null)))
+  }
+
+  @Test
+  def testArrayHash(): Unit = {
+    // test primitive type
+    val hashFunc1 = HashCodeGenerator
+      .generateArrayHash(
+        new CodeGeneratorContext(new Configuration(), classLoader),
+        new IntType(),
+        "name")
+      .newInstance(classLoader)
+
+    val array1 = new GenericArrayData(Array(1, 5, 7))
+    assertEquals(1123, hashFunc1.hashCode(array1))
+
+    // test complex map type of element
+    val hashFunc2 = HashCodeGenerator
+      .generateArrayHash(
+        new CodeGeneratorContext(new Configuration(), classLoader),
+        new MapType(new IntType(), new VarCharType()),
+        "name")
+      .newInstance(classLoader)
+
+    val mapData = new GenericMapData(
+      Map(1 -> StringData.fromString("ron"), 5 -> StringData.fromString("danny"), 10 -> null))
+    val array2 = new GenericArrayData(Array[AnyRef](mapData))
+    assertEquals(93178751, hashFunc2.hashCode(array2))
+
+    // test complex row type of element
+    val hashFunc3 = HashCodeGenerator
+      .generateArrayHash(
+        new CodeGeneratorContext(new Configuration(), classLoader),
+        RowType.of(new IntType(), new BigIntType()),
+        "name")
+      .newInstance(classLoader)
+
+    val array3 = new GenericArrayData(
+      Array[AnyRef](GenericRowData.of(ji(5), jl(8)), GenericRowData.of(ji(25), jl(52))))
+    assertEquals(14520, hashFunc3.hashCode(array3))
+
+    // test hash code for RowData
+    CommonTestUtils.assertThrows(
+      "ArrayData hash function doesn't support to generate hash code for RowData.",
+      classOf[RuntimeException],
+      () => hashFunc3.hashCode(GenericRowData.of(null))
+    )
+
+    // test hash code for MapData
+    CommonTestUtils.assertThrows(
+      "ArrayData hash function doesn't support to generate hash code for MapData.",
+      classOf[RuntimeException],
+      () => hashFunc3.hashCode(new GenericMapData(null))
+    )
+  }
+
+  @Test
+  def testMapHash(): Unit = {
+    // test primitive type
+    val hashFunc1 = HashCodeGenerator
+      .generateMapHash(
+        new CodeGeneratorContext(new Configuration(), classLoader),
+        new IntType(),
+        new VarCharType(),
+        "name")
+      .newInstance(classLoader)
+
+    val map1 = new GenericMapData(
+      Map(1 -> StringData.fromString("ron"), 5 -> StringData.fromString("danny"), 10 -> null))
```

Comment:
Please test hashcode of 2 maps are equal, which have entries in a different order and contain null values.